### ISSUE TITLE
Color Swap Highlighter does not restore color, if (shared) material is modified at runtime

### ISSUE THREAD

**Issue by burnedikt** [thestonefox/VRTK#844]
Happens if you have gameobjects / interactibles which dynamically change their color at runtime. 

E.g. a random color is assigned to a cube in the `Start()` method using `material.color = myColor`). If afterwards a controller touches the cube, it will correctly adopt the selected highlightColor but will never go back to its initial state. 

The reason is that on the first occasion that the material of a renderer is modified, it is cloned (if it is also used by other renderers) and from then on, this copy is being modified. As such, the [`StoreOriginalMaterials` method](https://github.com/thestonefox/VRTK/blob/master/Assets/VRTK/Scripts/Interactions/Highlighters/VRTK_MaterialColorSwapHighlighter.cs#L113) already gets a reference to the cloned material and the [`ChangeToHighlightColor` method](https://github.com/thestonefox/VRTK/blob/master/Assets/VRTK/Scripts/Interactions/Highlighters/VRTK_MaterialColorSwapHighlighter.cs#L156) will modify the same instance, thus overwriting the initial state.

The solution would be not to backup the references to materials but to make actual copies (`new Material(oldMaterial)`).

**Comment by thestonefox** [comment]
If you make copies of the materials does that then not instance new materials which in turn increases draw calls because draw call batching can't happen then?

The point of the color swap highlighter is to not create new instances on the original material so if you have the same object 10 times in the scene with the same material, then it can batch those draw calls because they all share the same material.

If you clone each material on the object, then you have 10 new instances of that material and you can no longer draw call batch and the draw calls will go up by 10 times.

**Comment by burnedikt** [comment]
You're right about that. However, as soon as someone changes a property of the material, it is unique to that renderer anyways and it is no longer eligible for draw call batching. 

> Note:
This function automatically instantiates the materials and makes them unique to this renderer. It is your responsibility to destroy the materials when the game object is being destroyed. Resources.UnloadUnusedAssets also destroys the materials but it is usually only called when loading a new level. (https://docs.unity3d.com/ScriptReference/Renderer-material.html)

Basically, we need to find out if the material that we want to backup is already an instance (i.e. it has already been modified before) or if it is still in its initial state. To check that, we could compare the `materials[i]` with the `sharedMaterials[i]` and check for equality. If they are equal, we do not need to make a copy of the material, if they aren't we need to create a copy since otherwise we wil modify the parameters of the material instance directly without a way of going back.

**Comment by thestonefox** [comment]
The current solution only changes the material instance on highlighted material. The original material (because it is reset to the shared material) is still the original and not instanced and therefore is still eligible for draw call batching.

You can check this out yourself by adding the same object to the scene (say 5 times) and give each one a different highlight colour.

Each object will have the same default material and a new instance for the highlight material.

But when non of the objects are highlighted then they will all batch the draw calls.

**Comment by burnedikt** [comment]
Yes, I get that. But if you attach a script to the 5 objects which assigns a random color to the Material on `Start()` this will not work. The objects will stay highlighted forever after being touched once.

**Comment by thestonefox** [comment]
Yeah I know.

But the solution to that problem cannot break draw call batching.

So the potentials are either the existing highlighter is fixed in such a way where it doesn't break. Or a new highlighter is made that doesn't care about breaking draw call batching and it's up to the dev to choose.

If you're going to swap your material at runtime, then its fair to say you don't care about batching draw calls on that object imho.

**Comment by burnedikt** [comment]
That would be true if all the materials were swapped at runtime. But if an object (say a beach ball) has 2 different materials, one of which will be shared among all beach balls and the other will get a random color at runtime to distinguish the beach balls. You spawn 200 beach balls because you want to have some fun, so assigning the random color manually in the editor is no option. 

In this case, I'd say one would still care about batching the draw calls for the shared material. 

However, can you elaborate a bit because I am having trouble understanding why the current code even works in the first place: 

- To my best understanding, as soon as `renderer.materials` is called, the materials are all instantiated so [backing them up like this](https://github.com/thestonefox/VRTK/blob/master/Assets/VRTK/Scripts/Interactions/Highlighters/VRTK_MaterialColorSwapHighlighter.cs#L113) already backs up the instances ...
- ... and [restoring them](https://github.com/thestonefox/VRTK/blob/master/Assets/VRTK/Scripts/Interactions/Highlighters/VRTK_MaterialColorSwapHighlighter.cs#L100) should thus restore those instances but as I can see in the editor instead it seems to restore the shared materials. How?
- What's the use of setting the [`sharedMaterials` to the just backed up `sharedMaterials`](https://github.com/thestonefox/VRTK/blob/master/Assets/VRTK/Scripts/Interactions/Highlighters/VRTK_MaterialColorSwapHighlighter.cs#L114)?

Sorry, quite confused right now.

**Comment by thestonefox** [comment]
Apparently it's due to a quirk in Unity that as soon as you touch an instantiated material it automatically instances a new one, but if you then immediately after reset the shared material then Unity doesn't actually bother doing the instantiation (maybe that check happens at the end of the frame and if you've switched straight back before that's happened then you can touch the material without instancing a new one).

**Comment by thestonefox** [comment]
There may be another solution. Someone did a PR a while back using MPB to do the colour swapping and apparently that way doesn't break batching.

I never got chance to look at the PR but it may be a solution?

https://github.com/thestonefox/VRTK/pull/704

**Comment by burnedikt** [comment]
Will check it out.

**Comment by thestonefox** [comment]
Also, to replicate the issue that's being discussed (for newcomers!)

 * Go to Example `040_Controls_Panel_Menu`
 * Set the Flying Saucer interactable object to have a highlight colour.
 * Grab the object and change it's colour with the panel menu
 * Drop the object.

Expected: it should stay as the new colour
Actual: It goes back to the highlight colour even when not being touched.

**Comment by burnedikt** [comment]
Update: The Solution based on MaterialPropertyBlock highlighting worked fine for my use case. However, the script at https://github.com/thestonefox/VRTK/pull/704  needs to be updated to work with the most recent version of VRTK. 

Will update the PR so that other use cases like `040_Controls_Panel_Menu` can also be tested.

**Comment by burnedikt** [comment]
Updated https://github.com/thestonefox/VRTK/pull/845 to incorporate the MaterialPropertyBlock Highlighter initially implemented by [fight4dream](https://github.com/fight4dream). Works for the example posted above relating to `040_Controls_Panel_Menu`

**Comment by mograph** [comment]
I applied the fix from earlier in this thread but I'm still hitting the issue. Does the fact that I'm using drop zones change anything here? I also have multiple interactable objects nested within another. 

Essentially I've built an experience where the user builds a model. There are a bunch of GameObject that the user picks up and snaps together (like lego blocks). When placing some of these objects, the highlight color gets stuck.

The issue is intermittent...
